pkg/actuators/machine/reconciler.go

```diff
@@ -175,9 +194,120 @@ func (r *Reconciler) reconcileMachineWithCloudState(conditionFailed *ibmcloudpro
+			}
+			klog.Infof("%s: setting machine's status to Provisioning for replacement machine", r.machine.Name)
+			r.machine.Status.Phase = stringPointerRef(phaseProvisioning)
+			return nil
+		}
 		return fmt.Errorf("get instance failed with an error: %q", err)
 	}
 
+	// Check whether the machine is being deleted due to a replacement request and then determine if that deletion has passed a deadline, 'machineDeletedDeadline'
+	if newInstance.Status != nil && *newInstance.Status == statusDeleting {
+		if machineReplacementCondition := conditionTypeCheck(r.providerStatus.Conditions, ibmcloudproviderv1.MachineReplacement); machineReplacementCondition != nil {
+			// Update machine replacement status to in progress if it hasn't been yet
+			if machineReplacementCondition.Reason == ibmcloudproviderv1.MachineReplacementRequested {
+				klog.Infof("%s: machine replacement now in progress", r.machine.Name)
+				r.providerStatus.Conditions = reconcileProviderConditions(r.providerStatus.Conditions, ibmcloudproviderv1.IBMCloudMachineProviderCondition{
+					Type:    ibmcloudproviderv1.MachineReplacement,
+					Reason:  ibmcloudproviderv1.MachineReplacementInProgress,
+					Message: machineReplacementInProgressMessageCondition,
+					Status:  apicorev1.ConditionTrue,
+				})
+				return nil
+			} else if machineReplacementCondition.Reason == ibmcloudproviderv1.MachineReplacementFailed {
+				// If replacement failed, attempt to update the status to Failed, if not already done
+				klog.Infof("%s: machine replacement failed", r.machine.Name)
+				if r.machine.Status.Phase != nil && *r.machine.Status.Phase != phaseFailed {
+					r.machine.Status.Phase = stringPointerRef(phaseFailed)
+					return nil
+				}
+			}
+			// Machine replacement should be in progress
+			// Check whether the machine is past the deletion deadline, returning an error if so
+			klog.Infof("%s: checking if machine is past replacement deletion deadline", r.machine.Name)
+			deletePastDeadline, err := r.checkMachineDeadline(machineDeleteDeadline, machineReplacementCondition.LastTransitionTime.Format(time.RFC3339))
+			if err != nil {
+				return fmt.Errorf("failed calculating delete deadline with error: %q", err)
+			} else if deletePastDeadline {
+				// If machine has failed to be replaced by deadline, update status and return an error
+				klog.Warningf("%s: machine failed to delete for replacement within deadline", r.machine.Name)
+				r.providerStatus.Conditions = reconcileProviderConditions(r.providerStatus.Conditions, ibmcloudproviderv1.IBMCloudMachineProviderCondition{
+					Type:    ibmcloudproviderv1.MachineReplacement,
+					Reason:  ibmcloudproviderv1.MachineReplacementFailed,
+					Message: machineReplacementFailedMessageCondition,
+					Status:  apicorev1.ConditionTrue,
+				})
+				return machinecontroller.DeleteMachine("%s: machine failed to delete within %d minute deadline for replacement", r.machine.Name, machineDeleteDeadline)
```

Comment:
is it appropriate to move the machine into failed phase here?

Reply:
So I considered doing it right here, but, in case we have an issue performing the update to the Machine, with the new `MachineReplacementFailed` status, I wanted a process to get back here, to try updating the `MachineReplacementFailed` status again, and once that status is set, we can attempt (and re-attempt if necessary) to set the machine to `Failed`, just above
https://github.com/openshift/machine-api-provider-ibmcloud/pull/10/files/2ef626eb5c39b1c0a463d4a605f7e3567a53dab1#diff-ea007161ea89a6c6d90fbb45625ef6c1a02f78d6afca539c33064c91711cba54R227-R233

This is similar to the logic I have to mark machine's for replacement, where there are changes the Statuses we set and/or values we clear and try to delete the machine, any one of those attempts can fail (the `PatchMachine` can fail due to timeout), and without the assumption to retry, would leave the Status or values in bad states and break this entire cycle.
https://github.com/openshift/machine-api-provider-ibmcloud/pull/10/files/2ef626eb5c39b1c0a463d4a605f7e3567a53dab1#diff-ea007161ea89a6c6d90fbb45625ef6c1a02f78d6afca539c33064c91711cba54R271-R306

I could perform the same `PatchMachine` check here as well, but figured we could get away with simply having the `Reason == MachineReplacementFailed` check above to re-attempt the `Failed` Status change as many times as we need (if it wasn't already `Failed`)

Reply:
that makes sense to me, thanks for the explanation

Reply:
@JoelSpeed , ptal at this comment. @cjschaef and i talked about this in slack a little too and i'm ok with this, but my concern would be that we enter some sort of loop that never exits. that said, i'm happy with the explanation here.

Reply:
My concern in these blocks of code, is when we end up calling `PatchMachine` and it times out, and we lose our local changes, and we cannot repeat certain operations or update those statuses again.
```
E1102 19:42:13.950988       1 machine_scope.go:162] Failed to patch machine "eu-de-ocpbugs-1327-d4-rrrq4-worker-2-6s4wz": Timeout: request did not complete within requested timeout - context deadline exceeded
```

Reply:
I think this is ok, the code to me, if I've understood correctly is re-entrant if we fail to patch the machine. Patching partial bits of information on the status like this to make sure they persisted and then patching another part when we have observed that change is persisted is a pattern we use in other places as well.

My observation here is we set something in the IBM status saying we couldn't delete the Machine, and then observe that later to set the Machine failed, is that correct?

Reply:
That is correct, we set the Status as `ReplacementFailed` in the first iteration, and then set `Failed` in the followup iteration. To make sure we persist the data as we go, and hopefully retry those changes if a failure occurs trying to save those changse.